molgenis-data-security/src/main/java/org/molgenis/data/security/permission/PermissionServiceImpl.java

```diff
@@ -0,0 +1,387 @@
+          permission.getPermission().name(), objectIdentity.getType());
+    }
+    Sid sid = permission.getSid();
+    Set<LabelledPermission> current =
+        getPermissionsForObject(objectIdentity, singleton(sid), false);
+    if (current.isEmpty()) {
+      throw new UnknownAceException(objectIdentity, sid, "update");
+    }
+    deleteAce(sid, acl);
+    acl.insertAce(acl.getEntries().size(), permission.getPermission(), sid, true);
+    mutableAclService.updateAcl(acl);
+  }
+
+  @Override
+  @Transactional
+  public void updatePermissions(Set<Permission> permissions) {
+    for (Permission permission : permissions) {
+      entityHelper.checkEntityExists(permission.getObjectIdentity());
+      updatePermission(permission);
+    }
+  }
+
+  @Override
+  @Transactional
+  public void deletePermission(Sid sid, ObjectIdentity objectIdentity) {
+    entityHelper.checkEntityExists(objectIdentity);
+    MutableAcl acl = (MutableAcl) mutableAclService.readAclById(objectIdentity, singletonList(sid));
+    if (acl == null) {
+      throw new UnknownAceException(objectIdentity, sid, "delete");
+    }
+    deleteAce(sid, acl);
+  }
+
+  @Override
+  @Transactional
+  public void addType(String typeId) {
+    entityHelper.checkEntityTypeExists(typeId);
+    EntityType entityType = dataService.getEntityType(entityHelper.getEntityTypeIdFromType(typeId));
+    if (!mutableAclClassService.getAclClassTypes().contains(typeId)) {
+      mutableAclClassService.createAclClass(typeId, EntityIdentityUtils.toIdType(entityType));
+      // Create ACL's for existing rows
+      dataService
+          .findAll(entityType.getId())
+          .forEach(
+              entity -> {
+                try {
+                  mutableAclService.createAcl(new EntityIdentity(entity));
+                } catch (AlreadyExistsException e) {
+                  LOG.warn(
+                      "Acl for entity '{}' of type '{}' already exists",
+                      entity.getIdValue(),
+                      entityType.getIdValue());
+                }
+              });
+    } else {
+      throw new AclClassAlreadyExistsException(typeId);
+    }
+  }
+
+  @Override
+  @Transactional
+  public void deleteType(String typeId) {
+    entityHelper.checkEntityTypeExists(typeId);
+    mutableAclClassService.deleteAclClass(typeId);
+  }
+
+  @Override
+  public boolean exists(ObjectIdentity objectIdentity, Sid sid) {
+    MutableAcl acl = (MutableAcl) mutableAclService.readAclById(objectIdentity, singletonList(sid));
+    return acl.getEntries().stream().anyMatch(ace -> ace.getSid().equals(sid));
+  }
+
+  private void deleteAce(Sid sid, MutableAcl acl) {
+    int nrEntries = acl.getEntries().size();
+    for (int i = nrEntries - 1; i >= 0; i--) {
+      AccessControlEntry accessControlEntry = acl.getEntries().get(i);
+      if (accessControlEntry.getSid().equals(sid)) {
+        acl.deleteAce(i);
+        mutableAclService.updateAcl(acl);
```

Comment:
performance:  update he ACL once after updating the ACE list instead of for each ACE